### ISSUE TITLE
Improve BackendApplicationClient class

### ISSUE THREAD

**Issue by DiegoQueiroz** [oauthlib/oauthlib#603]
**Describe the feature**

In `BackendApplicationClient` class, the `client_id`, `scope`, and `**kwargs` parameters passed to the constructor are useless. Notice that even `client_id` being mandatory, it is not used at all (it is only stored). The `**kwargs` parameter, on the other hand, is completely ignored.

The `prepare_request_body` method overrides `scope` and ignores the one passed to the constructor. The same happens to `**kwargs`. This forces me to pass `client_id`/`scope`/`**kwargs` twice (actually, only once, since I can pass anything to the constructor, the result will be the same).

**Additional context**

My suggestion is to override the constructor in the `BackendApplicationClient` subclass with this code (or something similar):

```
from __future__ import absolute_import, unicode_literals

from ..parameters import parse_token_response, prepare_token_request
from oauthlib.common import generate_token
from .base import Client, AUTH_HEADER

class BackendApplicationClient(Client):
    def __init__(self, client_id,
                 default_token_placement=AUTH_HEADER,
                 token_type='Bearer',
                 access_token=None,
                 refresh_token=None,
                 mac_key=None,
                 mac_algorithm=None,
                 token=None,
                 scope=None,
                 state=None,
                 redirect_url=None,
                 state_generator=generate_token,
                 **kwargs):
        # forward parameters to super class
        super(BackendApplicationClient, self).__init__(client_id,
                 default_token_placement,
                 token_type,
                 access_token,
                 refresh_token,
                 mac_key,
                 mac_algorithm,
                 token,
                 scope,
                 state,
                 redirect_url,
                 state_generator,
                 **kwargs)
        # store kwargs (client_id and scope are already stored by super class method)
        # better still would be to add this piece of code directly in the `Client` superclass
        self.kwargs = kwargs

    def prepare_request_body(self, body='', scope=None, **kwargs):
        if scope is None:
            # no new scope provided, use the class scope
            scope = self.scope
        if not kwargs:
            # empty kwargs, use class kwargs
            kwargs = self.kwargs
        if self.client_id and 'client_id' not in kwargs:
            # kwargs does not contains client_id but the class has one, add it
            kwargs['client_id'] = self.client_id
        return prepare_token_request('client_credentials', body=body,
                                     scope=scope, **kwargs)
```

This way, the parameters passed to the constructor are used for something, while we keep compatibility with existing code.

**Comment by JonathanHuot** [comment]
Hi! Yes you're right, `kwargs` in constructor is not used. However the way `client_id` is handled has a bit changed recently (see https://github.com/oauthlib/oauthlib/pull/593).

Had you got a chance to look at master and see if it fits your needs ?

**Comment by DiegoQueiroz** [comment]
You were right, I hadn't look at master. My fault.

I found the recent changes fits my needs, but they still have problems, particularly in `BackendApplicationClient`.

In other classes `client_id` was deprecated (eg. `WebApplicationClient`). This is good, but inconsistent with other classes (eg. `BackendApplicationClient`). Also, `include_client_id` is ignored.

I don't know if there is a rational behind this, but it seems that `BackendApplicationClient` may follow `WebApplicationClient` behavior. Maybe the intention was that `BackendApp...` be a subclass of `WebApp..`?

In addition, the same treatment given to `code` in `WebApplicationClient` could be given to `scope` in `BackendApplicationClient` (I mean `scope = scope or self.scope`).

`**kwargs` is less problematic, but I feel it could be forwarded as well (maybe with include_kwargs, or something).

**Comment by JonathanHuot** [comment]
Hi @DiegoQueiroz, thanks for the research! a PR increasing the consistency between classes would be greatly appreciated !

Adding a new arg/parameter (e.g. include_kwargs) seems too overkill/or unnecessary, though.

**Comment by JonathanHuot** [comment]
I think the immediate concern was fixed in #682 and the handling of `kwargs` is left. However the `kwargs` applies to other part of the code, not only `BackendApplicationClient`.

If someone would like to start a PR/issue with an API changes proposal regarding the `kwargs`, please do so ! :) 

Thanks
